src/www/app/server/access_config.ts

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {makeConfig, SHADOWSOCKS_URI, SIP002_URI} from 'ShadowsocksConfig';
+import {ServerUrlInvalid, ServerIncompatible, ShadowsocksUnsupportedCipher} from '../../model/errors';
+import {ShadowsocksConfig} from '../config';
+
+export enum OutlineServerSupportedCipher {
+  CHACHA20_IETF_POLY1305 = 'chacha20-ietf-poly1305',
+  AES_128_GCM = 'aes-128-gcm',
+  AES_192_GCM = 'aes-192-gcm',
+  AES_256_GCM = 'aes-256-gcm',
+}
+
+export enum OutlineServerAccessType {
+  SHADOWSOCKS_URI,
+  SHADOWSOCKS_CONFIG,
+}
+
+export class OutlineServerAccessConfig implements ShadowsocksConfig {
+  type: OutlineServerAccessType;
+  isOutlineServer = false;
+
+  private _rawUri?: string;
+  private _host?: string;
+  private _port?: number;
+  private _method?: OutlineServerSupportedCipher;
+  private _password?: string;
+  private _name?: string;
+
+  constructor(accessData: string | ShadowsocksConfig) {
+    if (typeof accessData === 'string') {
+      this.fromUri(accessData);
```

Comment:
Constructors should not have to know about serialization formats. It prevents testing and reuse, and limits how it can be created.
Revert to having a factory function that does the deserialization.

Reply:
I had it as a static method earlier in the pr so I can just switch back to that

Reply:
It just got annoying when i had to `fromX().toString()` in order to validate it, that felt silly

Reply:
I suppose that could be wrapped in a class method as well!